src/atomEffect.ts

```diff
@@ -49,8 +48,9 @@ export function atomEffect(
       }
       ++ref.inProgress
       try {
-        await ref.cleanup?.()
-        ref.cleanup = await effectFn(get, setSelf as Setter)
+        await void 0
```

Comment:
Having this is safer, but I wonder if we can remove `async` at all. It has more control on users.
The caveat is that they may get "setSelf" warning on dev. But, if we document it well, maybe worth it.

Or, does any of tests fail if we remove `async` at L43?

Reply:
If removed, the setSelf will be somehow in sync function if you use not properly

Reply:
I think this is still async because of the `await void 0`. Otherwise the tests would fail as they depend on atomEffects setting values.

`setSelf` is either:
 (a) warn in dev mode if sync
 (b) set a value if !async
 (c) do nothing in prod mode if sync
 
The only way to set a value is if !async

```js
setSelf = (...args) => {
  if (import.meta.env?.MODE !== 'production' && isSync) {
    console.warn('setSelf function cannot be called in sync')
  }
  if (!isSync) {
    return writeAtom(atom, ...args)
  }
}
```

Reply:
Hmm, if we want this, it's required.
```js
atomEffect((get, set) => {
  const count = get(countAtom)
  set(doubleAtom, count * 2)
})
```
But, wouldn't it be a bad practice, and we want to educate avoiding it even with `atomEffect`?

Reply:
> Having this is safer, but I wonder if we can remove `async` at all. It has more control on users. The caveat is that they may get "setSelf" warning on dev. But, if we document it well, maybe worth it.
> 
> Or, does any of tests fail if we remove `async` at L43?

We can make effectFn synchronous, but it runs in the next microtask.

The workaround for async is `transferrable`
https://codesandbox.io/s/epic-nash-pqkfvj

Reply:
Well, there's no use case in my mind, which requires no async, at the moment.
So, I'm fine with going a safer way.

Reply:
```js
atomEffect((get, set) => {
  const count = get(countAtom)
  set(doubleAtom, count * 2)
})
```

I think this is still ok because the entire effectFn runs in the next microtask, after [`isSync = false`](https://github.com/pmndrs/jotai/blob/main/src/vanilla/store.ts#L440)

Reply:
Yes, what I mean is if we **want** to warn such usage, because it's usually a bad idea in Jotai way.

Reply:
On another thought, `async` is actually better, because it's likely that we call `set` in cleanupFn, and then invoke the next effectFn, ensuring the sequence.

Reply:
What if we instead queue up overlapping calls. So that inProgress only blocks on synchronus updates. I think this would be incrementally better than blocking entirely on async calls, but it would still be complicated and could allow concurrent setups without first tearing down the previous run.

Reply:
Sorry for the confusion. I just meant this PR is good to go.